Setuptools rust hook for cross compilation

Cynerd opened the pull request:
###### Description of changes
Introduces setup hook for setuptools-rust that provides environment variables to really perform cross compilation of Rust code when using setuptools-rust.

This is a followup of https://github.com/NixOS/nixpkgs/pull/208401 and https://github.com/NixOS/nixpkgs/issues/205807.

###### Things done

Added setup hook. I also added `rustc` and `cargo` to the propagated dependencies and removed them from packages using `setuptools-rust`. This is for sure optional and thus can be dropped if you disagree with it.

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Cynerd committed:
setuptools-rust: fix cross compilation

The setuptools-rust requires some environment variables to really
perform cross build, otherwise it just builds for build platform.

This adds setup hook that introduces these environment variables.

There are three variables.
The PYO3_CROSS_LIB_DIR has to point to the target's Python library
directory. This has to be directory for the target not for the build or
host. We have to choose the correct target Python. I am unsure how to do
that simply in nixpkgs and this this implementations just delays this
and waits for the correct Python when package using this hook is build.
The CARGO_BUILD_TARGET triggers cross compilation in setuptools-rust.
This is simply the Rust target specification.
The CARGO_TARGET_*_LINKER variable should not be essentially required
but setuptools-rust probably mangles the Rust build environment somewhat
and that results to the missing linker. By explicitly specifying it
using the environment variable we force the correct linker.

SuperSandro2000 commented on pkgs/development/python-modules/setuptools-rust/default.nix:
```diff
@@ -21,7 +25,15 @@ buildPythonPackage rec {
   nativeBuildInputs = [ setuptools-scm ];
 
-  propagatedBuildInputs = [ semantic-version setuptools typing-extensions ];
+  propagatedBuildInputs = [ semantic-version setuptools typing-extensions rustc cargo ];
```
```suggestion
  propagatedNativeBuildInputs = [ rustc cargo ];

  propagatedBuildInputs = [ semantic-version setuptools typing-extensions ];
```

Cynerd replied:
Dropping due to the @FRidh comment.

SuperSandro2000 commented on pkgs/development/python-modules/setuptools-rust/default.nix:
```diff
@@ -21,7 +25,15 @@ buildPythonPackage rec {
+
+  setupHook = ./setuptools-rust-hook.sh;
+  pyLibPrefix = python.libPrefix;
+  cargoBuildTarget = "${targetPackages.rust.toRustTargetSpec targetPackages.stdenv.hostPlatform}";
```
```suggestion
  cargoBuildTarget = targetPackages.rust.toRustTargetSpec targetPackages.stdenv.hostPlatform;
```

Cynerd replied:
Yeah... the string only complicates it here.

SuperSandro2000 commented on pkgs/development/python-modules/setuptools-rust/setuptools-rust-hook.sh:
```diff
@@ -0,0 +1,27 @@
+setuptoolsRustSetup() {
+    # This can work only if rustPlatform.cargoSetupHook is also included
+    if ! command -v cargoSetupPostPatchHook >/dev/null; then
+        echo "ERROR: setuptools-rust has to be used alongside with rustPlatform.cargoSetupHook."
```
```suggestion
        echo "ERROR: setuptools-rust has to be used alongside with rustPlatform.cargoSetupHook!"
```

Cynerd replied:
🤷

SuperSandro2000 commented on pkgs/development/tools/poetry2nix/poetry2nix/overrides/default.nix:
```diff
@@ -214,7 +214,7 @@ lib.composeManyExtensions [
               (old.nativeBuildInputs or [ ])
                 ++ lib.optionals (lib.versionAtLeast old.version "4")
-                (with pkgs.rustPlatform; [ rust.rustc rust.cargo cargoSetupHook self.setuptools-rust ]);
+                [ pkgs.rustPlatform.cargoSetupHook self.setuptools-rust ];
```
This file has a bit strange workflow and it needs to be upstreamed, too. 

https://github.com/nix-community/poetry2nix

Cynerd replied:
I take it that I won't touch it for now because I do not know much about it.

SuperSandro2000 commented on pkgs/development/python-modules/setuptools-rust/setuptools-rust-hook.sh:
```diff
@@ -0,0 +1,27 @@
+        exit 1
+    fi
+
+    # We can't easily decide which target Python interpreter should be used when
+    # building setuptools-rust. It is the same version as the one we use to
+    # build setuptools-rust for but possibly for a different target platform.
+    # The easiest way is to use Python as provided in the package that uses this
+    # hook.
+    local pypath
+    pypath="$(command -v python)"
+    pypath="${pypath%/bin/python}"
+    export PYO3_CROSS_LIB_DIR="$pypath/lib/@pyLibPrefix@"
```
Are you searching for python.pythonForBuild.interpreter ?

Cynerd replied:
Unfortunately not.  The `setuptools-rust` package is being build for build platform (host is build) with target to cross platform. The `pythonForBuild` gives me python for build. I need python for target platform.

FRidh replied:
We should make the spliced python available in the passthru probably, like we did now for build only.

Cynerd replied:
Actually by using hook that has correct python available solves the primary issue here.

FRidh reviewed (changes requested):
rustc and cargo were on purpose not propagated by the cargo setup hook so that the hook also functions with different rustc cargo versions. Thus this needs to be undone.

The setup hooks are also defined separately from the package. This is for a similar reason; users should be able to use the package without having the hook running.

Cynerd commented:
> rustc and cargo were on purpose not propagated by the cargo setup hook so that the hook also functions with different rustc cargo versions. Thus this needs to be undone.

That is something I asked for but got no response. A drop of the package changes is easy. Updating the packages will be more complicated.

> The setup hooks are also defined separately from the package. This is for a similar reason; users should be able to use the package without having the hook running.

@FRidh and where they should be created actually? I found that python setup hooks are being created in `pkgs/development/interpreters/python/hooks/default.nix`. Should I move hook there?

Cynerd committed:
python-modules/cryptography: fix cross-build

The rustPlatform.rustc has both host and target set to build platform.
The rustc has build and host platform same but target is correctly set
for cross compilation.

FRidh commented:
> @FRidh and where they should be created actually? I found that python setup hooks are being created in pkgs/development/interpreters/python/hooks/default.nix. Should I move hook there?

Yes, indeed there.

Cynerd commented:
It took me some time to get back to it, but now it contains the hook, hopefully in the correct location, and it also fixes the cryptography build. I left out the fix of other packages this time.

SuperSandro2000 commented on pkgs/development/interpreters/python/hooks/setuptools-rust-hook.sh:
```diff
@@ -0,0 +1,18 @@
+
+setuptoolsRustSetup() {
+    # This can work only if rustPlatform.cargoSetupHook is also included
+    if ! command -v cargoSetupPostPatchHook >/dev/null; then
```
```suggestion
    if ! type cargoSetupPostPatchHook 2>/dev/null >/dev/null; then
```
command only really works correct with entries in PATH and has rather weird behaviors with shell functions. type is a bit more robust

Cynerd replied:
Good point. Applied with a small tweak to the shell.

FRidh commented on pkgs/development/interpreters/python/hooks/default.nix:
```diff
@@ -160,6 +160,20 @@ in {
       };
     } ./setuptools-check-hook.sh) {};
 
+    setuptoolsRustBuildHook = callPackage ({ makePythonHook, setuptools-rust, rust }:
```
The hook cannot depend on a rust version. How would we deal with when we need to build against a different rust version?

Artturin replied:
doesn't seem like a issue here, it's just using rust for the functions

Cynerd replied:
@Artturin is fully correct here. It really only uses functions there. In the end, only required function is `toRustTargetSpec`.

@FRidh as far as I can see https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/rust/lib/default.nix does not depend on Rust version.

FRidh reviewed (changes requested):
Either the hook needs to be in `python/hooks`, and it can depend on a python package but not on any rust/cargo, or it needs to be in `rust/hooks`, and it can depend on a rust/cargo version but not on a python package.

FRidh commented:
And these are not rust version dependent?

On Fri, 14 Apr 2023, 18:18 Artturi, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pkgs/development/interpreters/python/hooks/default.nix
> <https://github.com/NixOS/nixpkgs/pull/212795#discussion_r1167040449>:
>
> > @@ -160,6 +160,20 @@ in {
>        };
>      } ./setuptools-check-hook.sh) {};
>
> +    setuptoolsRustBuildHook = callPackage ({ makePythonHook, setuptools-rust, rust }:
>
> doesn't seem like a issue here, it's just using rust for the functions
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/212795#discussion_r1167040449>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQHZ3YZLH7LCXWTB7P5T2DXBF2GHANCNFSM6AAAAAAUHUUKXQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

SuperSandro2000 commented on pkgs/development/python-modules/cryptography/default.nix:
```diff
@@ -50,8 +52,10 @@ buildPythonPackage rec {
     cffi
   ] ++ [
     rustPlatform.cargoSetupHook
```
I would expect setuptooks to go into nativeBuildInputs because they are run on the build platform and change settings for the compiler but I might be wrong here.

amjoseph-nixpkgs replied:
> I would expect setuptooks to go into nativeBuildInputs because they are run on the build platform

I think part of the problem is that `make-rust-platform.nix` isn't using `makeScopeWithSplicing`, so the splicing mechanism isn't working properly for it.  This is one of the reasons I encourage people to avoid using `rustPlatform` whenever possible (as @Cynerd pointed out, sometimes you can't).

Personally I don't understand why we need a `rustPlatform`; it's very confusing.  If you "flatten" `rustPlatform` and `rustPlatform.rust` you get only seven attributes.  Why not just put all seven of them in `all-packages.nix` and eliminate `rustPlatform`?

Nixpkgs has a *lot* of subtle bugs when it comes to dealing with things (like `rustPlatform`) in `all-packages.nix` that are attrsets-of-packages instead of just packages.  In a perfect world we would fix those bugs, but I'm not sure that is feasible on a reasonable timeline.

amjoseph-nixpkgs replied:
It appears that you have applied this suggestion.

Cynerd commented on pkgs/development/interpreters/python/hooks/setuptools-rust-hook.sh:
```diff
@@ -0,0 +1,18 @@
+
+    export PYO3_CROSS_LIB_DIR="@pyLibDir@"
+    export CARGO_BUILD_TARGET=@cargoBuildTarget@
+    # TODO theoretically setting linker should not be required because it is
+    # already set in pkgs/build-support/rust/hooks/default.nix but build fails
+    # on missing linker without this.
+    export CARGO_TARGET_@cargoLinkerVar@_LINKER=@targetLinker@
```
I don't understand how not to use `rustPlatform`. How can I access the `cargoSetupHook` unless I am using `buildRustPackage` or through `rustPlatform`? I can't use `buildRustPackage` because this is a Python hook primarily.

amjoseph-nixpkgs replied:
> I don't understand how not to use `rustPlatform`.

Let me be more explicit:

> If you need this, it means splicing isn't working right. That's usually caused by referencing ~things~ **rustc and cargo** via `rustPlatform`. Don't do that. Use my next suggestion instead.

amjoseph-nixpkgs replied:
It appears that you have applied this suggestion.

Cynerd commented on pkgs/development/interpreters/python/hooks/default.nix:
```diff
@@ -160,6 +160,20 @@ in {
+        substitutions = {
+          pyLibDir = "${python}/lib/${python.libPrefix}";
+          cargoBuildTarget = rust.toRustTargetSpec stdenv.hostPlatform;
+          cargoLinkerVar = lib.toUpper (
+              builtins.replaceStrings ["-"] ["_"] (
+                rust.toRustTarget stdenv.hostPlatform));
+          targetLinker = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc";
```
Thank you. Really good point.

amjoseph-nixpkgs replied:
It appears that you have applied this suggestion.

amjoseph-nixpkgs reviewed (approved):
Built `python3Packages.cryptography` on:
- [x] `aarch64-linux` (cross from `x86_64-linux`)
- [x] `mips64el-linux-gnuabi64` (cross from `x86_64-linux`)
- [x] `powerpc64le-linux`
- [x] `x86_64-linux`

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/960

Artturin commented:
there's a conflict and also needs to target staging, check CONTRIBUTING.md for how to rebase

mweinelt commented on pkgs/development/python-modules/cryptography/default.nix:
```diff
@@ -49,11 +51,12 @@ buildPythonPackage rec {
+  ];
 
-  buildInputs = [ openssl ]
+  buildInputs = [ openssl rustPlatform.cargoSetupHook ]
```
```suggestion
  buildInputs = [
  	openssl
  	rustPlatform.cargoSetupHook
  ]
```

I might need a rationally why hooks are now buildInputs as opposed to nativeBuildInputs.

amjoseph-nixpkgs replied:
Agreed.

@Cynerd, for `rustPlatform.cargoSetupHook` you can probably harvest the explanation from this comment: https://github.com/NixOS/nixpkgs/pull/212795#discussion_r1169379568

amjoseph-nixpkgs commented:
@Cynerd, after resolving the merge conflict `python-cryptography` no longer cross-builds for me.  Perhaps I resolved the conflict incorrectly; in any case please double-check that `python-cryptography` cross-compiles after you resolve the conflict.

amjoseph-nixpkgs reviewed (dismissed):
@Cynerd, after resolving the merge conflict `python-cryptography` no longer cross-builds for me.  Perhaps I resolved the conflict incorrectly; in any case please double-check that `python-cryptography` cross-compiles after you resolve the conflict.

Cynerd commented:
@amjoseph-nixpkgs I just tested it, and it doesn't help my issue. The issue here is with cross-compilation from amd64 to 32-bit systems. Node requires bitness to match it seems, and for that reason, we need to build it not in an amd64 environment but instead in an x86 one.

I suspect that, especially on new Macs that do not have 32-bit arm support (if I remember correctly) it won't even be possible to cross-compile Node to armv7l.

Cynerd commented:
@amjoseph-nixpkgs let me see.. (sorry for disregarding this PR for some time).

Cynerd commented:
> there's a conflict and also needs to target staging, check CONTRIBUTING.md for how to rebase

Targeting now staging.

amjoseph-nixpkgs commented:
This no longer works after #245475

amjoseph-nixpkgs commented:
> This no longer works after #245475.

I have reverted https://github.com/NixOS/nixpkgs/pull/245639 (well, I will once OfBorg wakes up).

amjoseph-nixpkgs reviewed (approved):
LGTM.

@Cynerd, in the future, please use the "Commit suggestion" button when you implement the verbatim changes suggested by a reviewer.  Or if it's some kind of sharing-credit thing, at least make a separate commit with the requested change (once the PR is approved *then* you can squash).

When you implement peoples' suggestions by force-pushing a new commit with other stuff mixed in it is a hassle for the reviewer to figure out exactly what happened.

amjoseph-nixpkgs commented:
@FRidh I would like to merge this, because a lot of stuff is waiting on this PR or at least something equivalent to it.  Have your concerns been addressed?  If not, can you suggest a concrete change, or else perhaps wait for your concern to be dealt with in a follow-up PR?

amjoseph-nixpkgs commented:
@ofborg eval

rnhmjoj commented:
Is this supposed to have fixed cross compiling cryptography (https://github.com/NixOS/nixpkgs/issues/205807)?
I have just tried building `pkgsCross.aarch64-multiplatform.python3.pkgs.cryptography` and `pkgsCross.raspberryPi.python3.pkgs.cryptography` and both fails with:
```
    cargo:warning=In file included from /nix/store/2c7sgx69p6mmp76cvmi5j6c72dj76jj8-python3-3.10.12/include/python3.10/Python.h:50,
    cargo:warning=                 from /build/cryptography-41.0.2/src/rust/target/arm-unknown-linux-gnueabihf/release/build/cryptography-cffi-6fefc26573e7d979/out/_openssl.c:57:
    cargo:warning=/nix/store/2c7sgx69p6mmp76cvmi5j6c72dj76jj8-python3-3.10.12/include/python3.10/pyport.h:746:2: error: #error "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?)."
    cargo:warning=  746 | #error "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?)."
    cargo:warning=      |  ^~~~~
    exit status: 1
```

rnhmjoj commented:
My mistake: it's only failing on armv6.

kjeremy commented:
I'm seeing this on armv7 too:

```
warning: In file included from /nix/store/pzf6dnxg8gf04xazzjdwarm7s03cbrgz-python3-3.10.12/include/python3.10/Python.h:50,
warning:                  from /build/cryptography-41.0.3/src/rust/target/armv7-unknown-linux-gnueabihf/release/build/cryptography-cffi-1eecc4eb12ccf2f6/out/_opens>
warning: /nix/store/pzf6dnxg8gf04xazzjdwarm7s03cbrgz-python3-3.10.12/include/python3.10/pyport.h:746:2: error: #error "LONG_BIT definition appears wrong for platfo>
warning:   746 | #error "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?)."
warning:       |  ^~~~~

error: failed to run custom build command for `cryptography-cffi v0.1.0 (/build/cryptography-41.0.3/src/rust/cryptography-cffi)`

```

Artturin commented:
Create a issue.
